File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters.Direct/Session/Strategy.cs
// <copyright file="Object.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Direct
{
    using Meta;

    public sealed class Strategy : Adapters.Strategy
    {
        internal Strategy(Adapters.Workspace session, IClass @class, long id) : base(session, @class, id) => this.DatabaseState = new DatabaseState(this, session.DatabaseConnection.GetRecord(this.Id));

        internal Strategy(Adapters.Workspace session, Adapters.DatabaseRecord databaseRecord) : base(session, databaseRecord) => this.DatabaseState = new DatabaseState(this, databaseRecord);
    }
}

File: typescript/libs/system/workspace/adapters-json/src/lib/session/strategy.ts
import {
  Session,
  Strategy as SystemStrategy,
} from '@allors/system/workspace/adapters';
import { Class } from '@allors/system/workspace/meta';
import { DatabaseState } from './originstate/database-origin-state';
import { DatabaseRecord } from '../database/database-record';

export class Strategy extends SystemStrategy {
  constructor(
    public override session: Session,
    public override cls: Class,
    public override id: number
  ) {
    super(session, cls, id);

    this.DatabaseState = new DatabaseState(
      this.object,
      session.workspace.database.getRecord(this.id)
    );
  }

  static fromDatabaseRecord(session: Session, databaseRecord: DatabaseRecord) {
    return new Strategy(session, databaseRecord.cls, databaseRecord.id);
  }
}
